crates/sui-core/src/execution_cache/cached_version_map.rs

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::cmp::Ordering;
+use std::collections::VecDeque;
+
+use sui_types::base_types::SequenceNumber;
+
+/// CachedVersionMap is a map from version to value, with the additional contraints:
+/// - The key (SequenceNumber) must be monotonically increasing for each insert. If
+///   a key is inserted that is less than the previous key, it results in an assertion
+///   failure.
+/// - Similarly, only the item with the least key can be removed.
+/// - The intent of these constraints is to ensure that there are never gaps in the collection,
+///   so that membership in the map can be tested by comparing to both the highest and lowest
+///   (first and last) entries.
+#[derive(Debug)]
+pub struct CachedVersionMap<V> {
+    values: VecDeque<(SequenceNumber, V)>,
+}
+
+impl<V> Default for CachedVersionMap<V> {
+    fn default() -> Self {
+        Self {
+            values: VecDeque::new(),
+        }
+    }
+}
+
+impl<V> CachedVersionMap<V> {
+    pub fn is_empty(&self) -> bool {
+        self.values.is_empty()
+    }
+
+    pub fn insert(&mut self, version: SequenceNumber, value: V) {
+        if !self.values.is_empty() {
+            let back = self.values.back().unwrap().0;
+            assert!(
+                back < version,
+                "version must be monotonically increasing ({} < {})",
+                back,
+                version
+            );
+        }
+        self.values.push_back((version, value));
+    }
+
+    pub fn all_lt_or_eq_rev<'a>(
+        &'a self,
+        version: &'a SequenceNumber,
+    ) -> impl Iterator<Item = &'a (SequenceNumber, V)> {
+        self.values
+            .iter()
+            .rev()
+            .take_while(move |(v, _)| v <= version)
+    }
+
+    pub fn get(&self, version: &SequenceNumber) -> Option<&V> {
+        if self.values.is_empty() {
+            return None;
+        }
+
+        for (v, value) in self.values.iter().rev() {
+            match v.cmp(version) {
+                Ordering::Less => return None,
+                Ordering::Equal => return Some(value),
+                Ordering::Greater => (),
+            }
+        }
+
+        None
+    }
+
+    pub fn get_prior_to(&self, version: &SequenceNumber) -> Option<(SequenceNumber, &V)> {
+        for (v, value) in self.values.iter().rev() {
+            if v < version {
+                return Some((*v, value));
+            }
+        }
+
+        None
+    }
+
+    pub fn get_last(&self) -> Option<&(SequenceNumber, V)> {
+        self.values.back()
+    }
+
+    // pop items from the front of the map until the first item is >= version
+    pub fn truncate(&mut self, limit: usize) {
+        while self.values.len() > limit {
+            self.values.pop_front();
+        }
+    }
+
+    // remove the value if it is the first element in values. otherwise mark it
```

Comment:
What does the "otherwise mark it for removal" mean?
The assert is also interesting. Is this some kind of paranoid check?